crates/executor/src/blockchain_tree/mod.rs

```diff
@@ -0,0 +1,811 @@
+//! Implementation of [`BlockchainTree`]
+pub mod block_indices;
+pub mod chain;
+
+pub use chain::{Chain, ChainId, ForkBlock};
+
+use reth_db::{cursor::DbCursorRO, database::Database, tables, transaction::DbTx};
+use reth_interfaces::{consensus::Consensus, executor::Error as ExecError, Error};
+use reth_primitives::{BlockHash, BlockNumber, ChainSpec, SealedBlock, SealedBlockWithSenders};
+use reth_provider::{
+    ExecutorFactory, HeaderProvider, ShareableDatabase, StateProvider, StateProviderFactory,
+    Transaction,
+};
+use std::{
+    collections::{BTreeMap, HashMap},
+    sync::Arc,
+};
+
+use self::block_indices::BlockIndices;
+
+#[cfg_attr(doc, aquamarine::aquamarine)]
+/// Tree of chains and its identifications.
+///
+/// Mermaid flowchart represent all blocks that can appear in blockchain.
+/// Green blocks belong to canonical chain and are saved inside database table, they are our main
+/// chain. Pending blocks and sidechains are found in memory inside [`BlockchainTree`].
+/// Both pending and sidechains have same mechanisms only difference is when they got committed to
+/// database. For pending it is just append operation but for sidechains they need to move current
+/// canonical blocks to BlockchainTree flush sidechain to the database to become canonical chain.
+/// ```mermaid
+/// flowchart BT
+/// subgraph canonical chain
+/// CanonState:::state
+/// block0canon:::canon -->block1canon:::canon -->block2canon:::canon -->block3canon:::canon --> block4canon:::canon --> block5canon:::canon
+/// end
+/// block5canon --> block6pending:::pending
+/// block5canon --> block67pending:::pending
+/// subgraph sidechain2
+/// S2State:::state
+/// block3canon --> block4s2:::sidechain --> block5s2:::sidechain
+/// end
+/// subgraph sidechain1
+/// S1State:::state
+/// block2canon --> block3s1:::sidechain --> block4s1:::sidechain --> block5s1:::sidechain --> block6s1:::sidechain
+/// end
+/// classDef state fill:#1882C4
+/// classDef canon fill:#8AC926
+/// classDef pending fill:#FFCA3A
+/// classDef sidechain fill:#FF595E
+/// ```
+///
+///
+/// main functions:
+/// * insert_block: insert block inside tree. Execute it and save it to database.
```

Comment:
Does this actually save to the database, or does `finalize_block` do that?

Reply:
The intention was initially to save it into `PendingBlock` table so that we don't lose them if we crash and recover. As you can imagine this is low priority and I removed this part, as it require some modification to `Compact` trait to save full block inside table. 

wrote better explanation there.

Reply:
and to add the `make_canonical` is the one that saves block to the database.